File: app/src/main/java/com/originb/inkwisenote2/modules/ocr/data/NoteTermFrequencyDao.java
package com.originb.inkwisenote2.modules.ocr.data;

import androidx.room.*;

import java.util.List;
import java.util.Set;

@Dao
public interface NoteTermFrequencyDao {
    @Query("SELECT * FROM note_term_frequency WHERE note_id = :noteIdToRead")
    List<NoteTermFrequency> readTermFrequenciesOfNote(Long noteIdToRead);

    @Query("SELECT * FROM note_term_frequency")
    List<NoteTermFrequency> getAllTermFrequencies();

    @Query("SELECT * FROM note_term_frequency WHERE term IN (:terms)")
    List<NoteTermFrequency> getNoteIdsForTerms(Set<String> terms);

    @Query("SELECT * FROM note_term_frequency WHERE term NOT IN (:terms)")
    List<NoteTermFrequency> getNoteIdsForAllTermsExcept(Set<String> terms);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertTermFrequenciesToDb(List<NoteTermFrequency> noteRelation);

    @Query("SELECT term, COUNT(*) as occurrenceCount FROM note_term_frequency WHERE term IN (:terms) GROUP BY term")
    List<TermOccurrence> getTermOccurrences(Set<String> terms);

    @Query("SELECT COUNT(DISTINCT note_id) as distinct_count FROM note_term_frequency")
    int getDistinctNoteIdCount();

    @Query("DELETE FROM note_term_frequency WHERE note_id = :noteId")
    void deleteTermFrequencies(Long noteId);
}

File: app/src/main/java/com/originb/inkwisenote2/modules/ocr/data/NoteTermFrequencyDao.kt
package com.originb.inkwisenote2.modules.ocr.data

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query

@Dao
interface NoteTermFrequencyDao {
    @Query("SELECT * FROM note_term_frequency WHERE note_id = :noteIdToRead")
    fun readTermFrequenciesOfNote(noteIdToRead: Long): MutableList<NoteTermFrequency>

    @JvmField
    @get:Query("SELECT * FROM note_term_frequency")
    val allTermFrequencies: MutableList<NoteTermFrequency>

    @Query("SELECT * FROM note_term_frequency WHERE term IN (:terms)")
    fun getNoteIdsForTerms(terms: MutableSet<String>): MutableList<NoteTermFrequency>

    @Query("SELECT * FROM note_term_frequency WHERE term NOT IN (:terms)")
    fun getNoteIdsForAllTermsExcept(terms: MutableSet<String>): MutableList<NoteTermFrequency>

    @Insert(onConflict = OnConflictStrategy.Companion.REPLACE)
    fun insertTermFrequenciesToDb(noteRelation: MutableList<NoteTermFrequency>)

    @Query("SELECT term, COUNT(*) as occurrenceCount FROM note_term_frequency WHERE term IN (:terms) GROUP BY term")
    fun getTermOccurrences(terms: MutableSet<String>): MutableList<TermOccurrence>

    @JvmField
    @get:Query("SELECT COUNT(DISTINCT note_id) as distinct_count FROM note_term_frequency")
    val distinctNoteIdCount: Int

    @Query("DELETE FROM note_term_frequency WHERE note_id = :noteId")
    fun deleteTermFrequencies(noteId: Long)
}
